File: answers/leetcode/Sqrt(x)/Sqrt(x).cpp
// binary search
//@result 1017 / 1017 test cases passed. Status: Accepted Runtime: 12 ms Submitted: 0 minutes ago You are here! Your runtime beats 7.64% of cpp submissions.

class Solution {
public:
    int mySqrt(int x) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		if (x == 0)
		{
			return x;
		}
		int left = 1, right = x;
		while (left <= right)
		{
			int mid = left + (right - left) / 2;
			int div = x / mid;
			if (mid < div)
			{
				left = mid + 1;
			}
			else if (mid > div)
			{
				right = mid - 1;
			} else
			{
				return mid;
			}
		}
		return right;
    }
};

File: answers/leetcode/Sqrt(x)/Sqrt(x).java
class Solution {
    public int mySqrt(int x) {
        // Avoid divided by 0
        if (0 == x) {
            return 0;
        }

        int left = 1;
        int right = x;
        while (left <= right) {
            int middle = (left + right) / 2;
            // Use division to accelerate
            if (middle <= x / middle) {
                if (middle <= x / middle && (middle + 1) > x / (middle + 1)) {
                    return middle;
                } else {
                    left = middle + 1;
                }
            } else {
                right = middle - 1;
            }
        }
        return -1;
    }
}